michbl/assignment1/assignment1.cpp

```diff
@@ -7,84 +7,129 @@ using namespace std;
+
+    int len1 = word1.length();
+    int len2 = word2.length();
     if (len1 != len2)
         return false;
+
     for (int i = 0; i < len1; i++) {
-        histogram[str1[i]]++;
-        histogram[str2[i]]--;
+        histogram[word1[i]]++;
+        histogram[word2[i]]--;
     }
+
     for (int i = 0; i < ALPHABET_SIZE; i++) {
-        if (caseSensitive || !isLetter((char)i)) {
-            if (histogram[i] != 0) {
+        char c = (char) i;
+        if (caseSensitive || !isLetter(c)) {
+            if (histogram[c] != 0) {
                 return false;
             }
         }
         else { // case insensitive and i is the ascii value of a letter
-            if (i >= 'a' && histogram[i] + histogram[i - 'a' + 'A'] != 0) {
-                return false;
-            }
-            else if (i < 'a' && histogram[i] + histogram[i - 'A' + 'a'] != 0) {
+            if (countCaseInsensitiveLettersInHistogram(c, histogram) != 0) {
                 return false;
             }
         }
     }
     return true;
 }
 
-//check whether 2 sentences (strings with/without whitespaces) are anagrams of each other
-bool anagrams(string str1, string str2, bool caseSensitive) {
-    vector<string> words1;
-    vector<string> words2;
-
-    //split str1 into words
-    size_t pos = 0, found = str1.find_first_of(' ', pos);
-    while (found != string::npos) {
-        words1.push_back(str1.substr(pos, found - pos));
-        pos = found+1;
-        found = str1.find_first_of(' ', pos);
-    }
-    words1.push_back(str1.substr(pos));
-
-    //split str2 into words
-    pos = 0;
-    found = str2.find_first_of(' ', pos);
-    while (found != string::npos) {
-        words2.push_back(str2.substr(pos, found - pos));
-        pos = found+1;
-        found = str2.find_first_of(' ', pos);
-    }
-    words2.push_back(str2.substr(pos));
+// check whether 2 sentences (strings with/without spaces) are anagrams of each other
+bool areSentencesAnagrams(string sentence1, string sentence2, bool caseSensitive) {
+    vector<string> words1 = splitSentenceToWords(sentence1);
+    vector<string> words2 = splitSentenceToWords(sentence2);
 
     if (words1.size() != words2.size()) {
         return false;
     }
 
-    bool match;
+    // try to match each word from sentence1 to a word from sentence2
     while (!words1.empty()) {
-        match = false;
         vector<string>::iterator it1 = words1.begin();
-        vector<string>::iterator it2 = words2.begin();
-        while (it2 != words2.end()) {
-            if (anagramsAux(*it1, *it2, caseSensitive)) {
-                match = true;
-                words1.erase(it1);
-                words2.erase(it2);
-                break;
-            }
-            it2++;
-        }
-        if (!match)
+        vector<string>::const_iterator it2 = findAnagramInList(words2, *it1, caseSensitive);
+        if (it2 == words2.end()) {
             return false;
+        }
+        else {
+            words1.erase(it1);

```

Comment:
Questions for thought:
What is the complexity of this algorithm? Can you do better?